File: backend/services/received_invoice_service.py
from db.core import get_session
from db.models import ReceivedInvoice, Invoice
from schemas.invoice_schema import ReceivedInvoiceSchema
from utils.api_response import APIResponse
from constants.constants import INVOICE_STATUS
import datetime


class ReceivedInvoiceService:
    @classmethod
    def create_received_invoice(cls, invoice: ReceivedInvoiceSchema):
        with get_session() as db:

            original_invoice = (
                db.query(Invoice).filter_by(po_number=invoice.po_number).first()
            )

            received_invoice = ReceivedInvoice(**invoice.dict())
            db.add(received_invoice)

            if original_invoice.balance is None:
                original_invoice.balance = original_invoice.amount

            if original_invoice.balance - invoice.amount >= 0:
                original_invoice.balance -= invoice.amount

            if original_invoice.balance == 0:
                original_invoice.status = INVOICE_STATUS["COMPLETED"]
                original_invoice.completed_date = datetime.datetime.now()

            db.add(original_invoice)
            db.commit()

            return APIResponse(
                success=True, message="received invoice recorded successfully"
            )

    @classmethod
    def get_received_invoices(cls):
        with get_session() as db:
            r_invoices = (
                db.query(ReceivedInvoice)
                .order_by(ReceivedInvoice.created_at.desc())
                .all()
            )

            return APIResponse(
                success=True, message="found received invoices", data=r_invoices
            )

    @classmethod
    def get_received_invoices_by_po(cls, po_number):
        with get_session() as db:
            invoices = (
                db.query(ReceivedInvoice)
                .filter_by(po_number=po_number)
                .order_by(ReceivedInvoice.created_at.desc())
                .all()
            )
            return APIResponse(success=True, message="fetched invoices", data=invoices)

File: backend/src/services/received-invoice-service.js
import { INVOICE_STATUS } from "../constants/constants.js";
import { Invoice, ReceivedInvoice } from "../db/models.js";

/*
  receiving an invoice moves money against the parent invoice (matched on PO)
  and closes it once the full quoted amount has been paid.
*/
export async function createReceivedInvoice(r_invoice, received_by = null) {
  const db_invoice = await Invoice.findOne({
    po_number: r_invoice.po_number,
    is_deleted: false,
  });

  if (!db_invoice) {
    return { received: null, message: "no invoice found for that PO number" };
  }

  const outstanding = db_invoice.amount - db_invoice.amount_paid;

  // invoice already paid cannot receive any further
  if (outstanding <= 0) {
    return { received: null, message: "invoice payment already cleared" };
  }

  if (r_invoice.amount > outstanding) {
    return {
      received: null,
      message: `amount exceeds the outstanding balance of ${outstanding}`,
    };
  }

  const db_received = await ReceivedInvoice.create({
    ...r_invoice,
    invoice: db_invoice._id,
    received_by,
  });

  db_invoice.amount_paid += r_invoice.amount;

  // check if invoice can be marked complete
  if (db_invoice.amount_paid >= db_invoice.amount) {
    db_invoice.status = INVOICE_STATUS.COMPLETED;
    db_invoice.completed_date = new Date();
  }

  await db_invoice.save();

  return { received: db_received, message: "invoice received successfully" };
}

export async function getReceivedInvoices(filter = {}) {
  return ReceivedInvoice.find({ is_deleted: false, ...filter }).sort({
    createdAt: -1,
  });
}

export async function getReceivedInvoicesByPO(po_number) {
  return getReceivedInvoices({ po_number });
}
